### ISSUE TITLE
MySQL 8 issue with "contains" selector operator and single quotes

### ISSUE THREAD

**Issue by chris-shorter** [processwire/processwire-issues#1950]
### Short description of the issue

When using the "contains" selector (e.g. `*=`) and a search term wrapped in opening and closing single quote characters of `‘` and `’`, e.g. `"‘aaa’"`, while using MySQL 8.0.x - a fatal error occurs.

MySQL 5.7 and Mariadb do not seem to have this problem

### Expected behavior

Standard page find is performed

### Actual behavior

The following error occurs:

`Error: Exception: SQLSTATE[HY000]: General error: 3854 Cannot convert string '(\b|[[...' from utf8mb3 to utf16le (in wire/core/PageFinder.php line 821)`

### Optional: Suggestion for a possible fix

Might not be the best method but we have found it is possible to fix this by changing the `rlikeValue` method of `DatabaseQuerySelectFulltext.php` to use multibyte safe preg_replace - however I'm not sure how desirable this is since it would add a new extension requirement to PW (albeit a common one) - e.g.

```php
// DatabaseQuerySelectFulltext.php lines 1270-1275
if(strpos($likeValue, "'") !== false || strpos($likeValue, "’") !== false) {
    // match either straight or curly apostrophe
    $likeValue = mb_ereg_replace('/[\'’]+/', '(\'|’)', $likeValue);
    // if word ends with apostrophe then apostrophe is optional
    $likeValue = rtrim(mb_ereg_replace("('|’) ", "('|’)? ", "$likeValue "));
}
```

I'm also not an expert php developer and haven't explored the wider implications of this or tested if it affects other selector operators.

### Steps to reproduce the issue

1. Start an environment running mysql 8.0. I have reproduced this on ddev and on a cPanel server
2. Perform a selector lookup for a term wrapped in single quote characters - this part is slightly inconsistent as `‘ipsum dolor’` will error but `‘ipsum’` will not, `‘aaa’` will error but `‘aaaa’` will not, however the following code should produce an error:

```php
<?php

    $searchterm = "‘ipsum dolor’";
    $results = $pages->find("title*=$searchterm");

    foreach ($results as $result) {
        echo $result->title . '<br>';
    }

?>
```

Sanitizing the search term with `selectorValue` doesn't make a difference.

### Setup/Environment

* ProcessWire version: 3.0.240 or 3.0.229
* MySQL version: 8.0.x

**Comment by adrianbj** [comment]
Hi @chris-shorter - it looks like you should be able to fix by changing the character set from utf8mb3 to utf8mb4.

It has been discussed here: https://stackoverflow.com/questions/72180043/incorrect-string-value-error-for-unconventional-characters

Do you have any particular reason for everything being in utf8mb3? Keep in mind that these days utf8mb4 is faster than utf8mb3.

**Comment by chris-shorter** [comment]
I didn't really opt for one or the other really, this is just what I had after a fresh install of PW on a default ddev environment

**Comment by matjazpotocnik** [comment]
@chris-shorter this fix https://github.com/processwire/processwire/commit/807e94e22a13146ac830de43500eb31d6e0d2803 make utf8mb4 and InnoDB the default settings for MySQL (rather than utf8 and MyISAM). Is this enough to close this issue?

**Comment by chris-shorter** [comment]
I ran a new install on mysql 8.0 from that commit and I still get the same issue when searching for a page with the same snippet

![image](https://github.com/user-attachments/assets/7cd282d0-1058-404f-9e06-23f03facbb7d)

**Comment by matjazpotocnik** [comment]
@chris-shorter, thank you for testing; it looks like the character set utf8mb3/utf8mb4 doesn't play a role in this issue. Will see if Ryan finds time to fix this.

**Comment by ryancramerdesign** [comment]
@chris-shorter @matjazpotocnik While I can't seem to duplicate the issue here, I can see from the query log that the `‘` character is getting mangled, so the error makes sense. We could convert the preg_replace() to use the `/u` flag, but this particular replacement doesn't seem to need a regex at all, so I've just converted it to a  multibyte safe str_replace(), and the character is no longer managed in the query log. Does it also fix it there?

**Comment by matjazpotocnik** [comment]
@chris-shorter, can you please confirm the fix?

**Comment by chris-shorter** [comment]
Yes I can confirm this fix stops the error and results are returned
